core/src/Dscp/Core/Foundation/Instances.hs

```diff
@@ -103,23 +313,210 @@ instance Serialise Timestamp where
+        ]
+    decode = do
+        decodeListLenOf 2
+        ptwTx      <- decode
+        ptwWitness <- decode
+        return $ PublicationTxWitnessed {..}
+
+instance Serialise GTx where
+    encode = \case
+        GMoneyTx tx -> mconcat
+            [ encodeListLen 2
+            , encodeWord 0
+            , encode tx
+            ]
+        GPublicationTx publicationTx -> mconcat
+            [ encodeListLen 2
+            , encodeWord 1
+            , encode publicationTx
+            ]
+    decode = do
+        decodeListLenOf 2
+        decodeWord >>= \case
+            0 -> GMoneyTx       <$> decode
+            1 -> GPublicationTx <$> decode
+            _ -> fail "Unexpected tag"
+
+instance Serialise GTxId where
+    encode (GTxId hGTx) = encode hGTx
+    decode = GTxId <$> decode
+
+instance Serialise GTxWitnessed where
+    encode = \case
+        GMoneyTxWitnessed txWitnessed -> mconcat
+            [ encodeListLen 2
+            , encodeWord 0
+            , encode txWitnessed
+            ]
+        GPublicationTxWitnessed publicationTxWitnessed -> mconcat
+            [ encodeListLen 2
+            , encodeWord 1
+            , encode publicationTxWitnessed
+            ]
+    decode = do
+        decodeListLenOf 2
+        decodeWord >>= \case
+            0 -> GMoneyTxWitnessed       <$> decode
+            1 -> GPublicationTxWitnessed <$> decode
+            _ -> fail "Unexpected tag"
+
+instance Serialise BlockToSign where
+    encode (BlockToSign difficulty slotId hHeader hBlockBody) = mconcat
+        [ encodeListLen 4
+        , encode difficulty
+        , encode slotId
+        , encode hHeader
+        , encode hBlockBody
+        ]
+    decode = do
+        decodeListLenOf 4
+        difficulty <- decode
+        slotId     <- decode
+        hHeader    <- decode
+        hBlockBody <- decode
+        return $ BlockToSign difficulty slotId hHeader hBlockBody
+
+instance Serialise SlotId where
+    encode (SlotId w) = encode w
+    decode = SlotId <$> decode
+
+instance Serialise Difficulty where
+    encode = encode . unDifficulty
+    decode = Difficulty <$> decode
+
+instance Serialise Header where
+    encode (Header {..}) = mconcat
+        [ encodeListLen 5
+        , encode hSignature
+        , encode hIssuer
+        , encode hDifficulty
+        , encode hSlotId
+        , encode hPrevHash
+        ]
+    decode = do
+        decodeListLenOf 5
+        hSignature  <- decode
+        hIssuer     <- decode
+        hDifficulty <- decode
+        hSlotId     <- decode
+        hPrevHash   <- decode
+        return $ Header {..}
+
+instance Serialise Block where
+    encode (Block {..}) = mconcat
+        [ encodeListLen 2
+        , encode bHeader
+        , encode bBody
+        ]
+    decode = do
+        decodeListLenOf 2
+        bHeader <- decode
+        bBody   <- decode
+        return $ Block {..}
+
+instance Serialise BlockBody where
+    encode = encode . bbTxs
+    decode = BlockBody <$> decode
```

Comment:
Thanks, that's a lot of work!

However, I personally think that we could go in a less boilerplate-y way. Except for long instances definition, now we have to write roundtrip tests.

Don't you mind thinking about how could we reduce manual work on defining these instances?
Out of the box idea: use `HList` or `Rec` as an intermediate representation for CBOR encoded list. `HList <-> Encoding` conversion can be applied automatically, while `type <-> HList` is still to be defined manually. Profit: you cannot mistake in specifying list length and it's pretty difficult to mistaken in arguments order. As a bonus, `vinyl` suggests a [function](http://hackage.haskell.org/package/vinyl-0.11.0/docs/Data-Vinyl-Curry.html#v:rcurry) which allows filling a constructor with values from `Rec`. I'm not sure that this approach lacks its downsides though.

There is a slight problem here, that no one yet understands what do we want from these instances and where flexibility is required :sweat_smile:. I hope you will find a solution )

Reply:
You are right, I started trying to see what the difficulties were and ended up making instances in one of the 3 ways described above in the description.
As you said there are a couple of problems with this approach (list length is very easy to mess up, sum-type "word" conversion is even more fragile), but also has a couple advantages (the decode side made with all the record fields spelled out will force one to update the definition when necessary).

I'll give a look at your suggestions, it may be possible to at least remove some of the boilerplate, this can only be a benefit, and pay attention at how easy can we modify them later.